lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -397,6 +397,26 @@ static bool shouldAllowReferenceToUnavailableInSwiftDeclaration(
   return false;
 }
 
+// Utility function to help determine if noasync diagnostics are still
+// appropriate even if a `DeclContext` returns `false` from `isAsyncContext()`.
+static bool shouldTreatDeclContextAsAsyncForDiagnostics(const DeclContext *DC) {
+  switch (DC->getContextKind()) {
+  case DeclContextKind::AbstractFunctionDecl: {
+    auto *FD = dyn_cast<FuncDecl>(DC->getAsDecl());
+    if (FD && FD->isDeferBody()) {
+      // If this is a defer body, we should delegate the determination
+      // to its parent.
+      return shouldTreatDeclContextAsAsyncForDiagnostics(DC->getParent());
+    } else {
+      return DC->isAsyncContext();
+    }
+  }
+
+  default:
+    return DC->isAsyncContext();
+  }
+}
```

Comment:
This is a matter of style and up to you as the author, but I'd prefer a more compact implementation since we only care about function decl contexts:
```suggestion
static bool shouldTreatDeclContextAsAsyncForDiagnostics(const DeclContext *DC) {
  if (auto *FD = dyn_cast<FuncDecl>(DC->getAsDecl())) {
    // If this is a defer body, we should delegate the determination
    // to its parent.
    if (FD->isDeferBody())
      return shouldTreatDeclContextAsAsyncForDiagnostics(DC->getParent());
  }
  return DC->isAsyncContext();
}
```

Reply:
thanks for the suggestion, that seems like a nicer formulation to me. i ran into some issues with things crashing using a similar approach b/c i _think_ `DC->getAsDecl()` may return null in some circumstances (function decl in top level code?) and `dyn_cast` doesn't like null arguments. will rework to something like this with an additional null check.